core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalSystem.java

```diff
@@ -137,6 +147,35 @@ public CatchupFuture catchup(Map<String, Long> journalSequenceNumbers) throws IO
     return CatchupFuture.allOf(futures);
   }
 
+  @Override
+  public void waitForCatchup() {
+    long start = System.currentTimeMillis();
+    try (Timer.Context ctx = MetricsSystem
+        .timer(MetricKey.MASTER_UFS_JOURNAL_CATCHUP_TIMER.getName()).time()) {
+      CommonUtils.waitFor("journal catch up to finish", () -> {
+        for (UfsJournal journal : mJournals.values()) {
+          UfsJournalCheckpointThread.CatchupState catchupState = journal.getCatchupState();
+          if (catchupState != UfsJournalCheckpointThread.CatchupState.CATCHUP_DONE) {
+            return false;
+          }
+        }
+        return true;
+      }, WaitForOptions.defaults().setTimeoutMs(
+          (int) ServerConfiguration.getMs(PropertyKey.MASTER_UFS_JOURNAL_MAX_INITIAL_REPLAY_TIME))
+          .setInterval(Constants.SECOND_MS));
+    } catch (InterruptedException | TimeoutException e) {
+      LOG.info("Journal catchup is interrupted or timeout", e);
+      if (mInitialCatchupTimeMs == -1) {
```

Comment:
Why do we need to check that we have not set the catchup time yet? Shouldn't the value reflect the latest catchup time?

Reply:
There are several catchups in the current implementation:
```
1. Initial catchup
2.start primary selector to join leader election
while {
    3. catchup when step back to become the secondary master and before checking if it's the primary
    mLeaderSelector.waitForState(State.PRIMARY);
    gainPrimacy()
    mLeaderSelector.waitForState(State.SECONDARY);
   losePrimacy();
}
```
I have one metric MASTER_UFS_JOURNAL_CATCHUP_TIMER to record the catchup time for all the catchup.
and have a MASTER_UFS_JOURNAL_INITIAL_REPLAY_TIME_MS to record the first initial catchup time.

The initial replay time should only be recorded once and not overridden by future catchups